Include/internal/pycore_critical_section.h

```diff
@@ -108,6 +109,25 @@ extern "C" {
         _PyCriticalSection2_End(&_cs2);                                 \
     }
 
+// Specialized version of critical section locking called to safely use
+// PySequence_Fast APIs under nogil
+// For performance, the argument *to* PySequence_Fast is provided to the
+// macro, not the *result* of PySequence_Fast (which would require an extra
+// test to determine if the lock must be held)
+# define Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST(original)              \
+    {                                                                   \
+        PyObject *_orig_seq = _PyObject_CAST(original);                 \
+        const bool _should_lock_cs = PyList_CheckExact(_orig_seq);      \
+        _PyCriticalSection _cs;                                         \
+        if (_should_lock_cs)                                            \
+            _PyCriticalSection_Begin(&_cs,                              \
+                                     &_orig_seq->ob_mutex)  \
```

Comment:
Let's write this with a bracketed if statement. It doesn't enforce the `;` in caller code, but I think that's okay.

```suggestion
        if (_should_lock_cs) {                                          \
            _PyCriticalSection_Begin(&_cs, &_orig_seq->ob_mutex);       \
        }
```

Reply:
Sounds reasonable. I avoided them to enforce the semi-colon, but that does mean potential weirdness with the `if` condition controlling more than expected with misuse of the comma, so it's not a strong preference (and as long as the GIL exists, it'll need the semi-colon to build in that mode anyway).